Nullable annotations for System.Security.Cryptography.Xml

SteveDunn opened the pull request:
Fixes #67197

SteveDunn committed:
First pass

SteveDunn committed:
More annotations

SteveDunn commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/KeyInfoX509Data.cs:
```diff
@@ -261,11 +261,13 @@ public override void LoadXml(XmlElement element!!)
-            XmlNodeList x509SubjectNameNodes = element.SelectNodes("ds:X509SubjectName", nsm);
-            XmlNodeList x509CertificateNodes = element.SelectNodes("ds:X509Certificate", nsm);
-            XmlNodeList x509CRLNodes = element.SelectNodes("ds:X509CRL", nsm);
+            //red flag - no existing null checks before usage, so should we assume that it's not null,
```
Should we assume that `SelectNodes` will _not_ return null? Or should we check for null? Or should we do something else in this case?

SteveDunn replied:
> Should we assume that `SelectNodes` will _not_ return null? Or should we check for null? Or should we do something else in this case?

krwq replied:
It's more important to get public APIs right than getting implementation details right. In this specific case select nodes can only return null if specific XmlNode you're running it on will return null when you create an XPathNavigator on it which I don't believe is a case here:
https://github.com/dotnet/runtime/blob/6a889d234267a4c96ed21d0e1660dce787d78a38/src/libraries/System.Private.Xml/src/System/Xml/Dom/XmlNode.cs#L93-L97

SteveDunn replied:
OK, thank you.

am11 commented on src/libraries/System.Security.Cryptography.Xml/tests/KeyInfoX509DataTest.cs:
```diff
@@ -61,7 +61,7 @@ public class KeyInfoX509DataTest
             0x9C,0x30,0x0C,0x06,0x03,0x55,0x1D,0x13,0x04,0x05,0x30,0x03,0x01,0x01,0xFF,0x30,0x09,0x06,0x07,0x2A,0x86,0x48,0xCE,0x38,0x04,0x03,0x03,0x2F,0x00,0x30,0x2C,0x02,0x14,0x5A,0x1B,0x2D,0x08,0x0E,0xE6,0x99,0x38,0x8F,0xB5,0x09,0xC9,0x89,0x79,0x7E,0x01,0x30,0xBD,0xCE,0xF0,0x02,0x14,0x71,0x7B,0x08,0x51,0x97,0xCE,0x4D,0x1F,0x6A,0x84,0x47,0x3A,0xC0,0xBD,0x13,0x89,0x81,0xB9,0x01,0x97 };
 
-        static byte[] x509crl = { 0x30, 0x82, 0x01, 0x05, 0x30, 0x72, 0x02, 0x01, 0x01, 0x30, 0x0B, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x05, 0x30, 0x51, 0x31, 0x0B, 0x30, 0x09, 0x06, 0x03, 0x55, 0x04, 0x06, 0x13, 0x02, 0x55, 0x53, 0x31, 0x18, 0x30, 0x16, 0x06, 0x03, 0x55, 0x04, 0x0A, 0x13, 0x0F, 0x55, 0x2E, 0x53, 0x2E, 0x20, 0x47, 0x6F, 0x76, 0x65, 0x72, 0x6E, 0x6D, 0x65, ...
+        static byte[]? x509crl = { 0x30, 0x82, 0x01, 0x05, 0x30, 0x72, 0x02, 0x01, 0x01, 0x30, 0x0B, 0x06, 0x09, 0x2A, 0x86, 0x48, 0x86, 0xF7, 0x0D, 0x01, 0x01, 0x05, 0x30, 0x51, 0x31, 0x0B, 0x30, 0x09, 0x06, 0x03, 0x55, 0x04, 0x06, 0x13, 0x02, 0x55, 0x53, 0x31, 0x18, 0x30, 0x16, 0x06, 0x03, 0x55, 0x04, 0x0A, 0x13, 0x0F, 0x55, 0x2E, 0x53, 0x2E, 0x20, 0x47, 0x6F, 0x76, 0x65, 0x72, 0x6E, 0x6D, 0x65 ...
```
Doe it break the compilation if this change is reverted? Test projects usually don't have nullable:enable.

SteveDunn replied:
Hi - I'm not sure how that ended up being changed! Well spotted.

SteveDunn committed:
Rever mistaken change to test

SteveDunn committed:
More annotations

maxkoshevoi commented:
If this is not ready for review yet, please make this PR an actual draft. There should be a button to do that:

![image](https://user-images.githubusercontent.com/6609929/160279875-40189bf1-0bc0-468a-ae65-05be34c3727e.png)

maxkoshevoi commented:
If this is not ready for review yet, please make this PR an actual draft. There should be a button to do that:

![image](https://user-images.githubusercontent.com/6609929/160279875-40189bf1-0bc0-468a-ae65-05be34c3727e.png)

SteveDunn committed:
More annotations

SteveDunn commented:
> If this is not ready for review yet, please make this PR an actual draft. There should be a button to do that:
> 
> ![image](https://user-images.githubusercontent.com/6609929/160279875-40189bf1-0bc0-468a-ae65-05be34c3727e.png)

Thanks @maxkoshevoi - I was wondering where that option was. Hope you're keeping well and safe!

maxkoshevoi commented:
> Thanks @maxkoshevoi - I was wondering where that option was. Hope you're keeping well and safe!

Thanks, I'm in relatively safe city, still have my job, roof over the head, and something to do in free time (walking around the city, and annotating runtime 😄), so things are quite good over all =)
Hope it'll be safe to return to my home in a month or so.

SteveDunn committed:
More annotations

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/C14NAncestralNamespaceContextManager.cs:
```diff
@@ -26,8 +26,8 @@ private void GetNamespaceToRender(string nsPrefix, SortedList attrListToRender,
             int rDepth;
-            XmlAttribute local = (XmlAttribute)nsLocallyDeclared[nsPrefix];
-            XmlAttribute rAncestral = GetNearestRenderedNamespaceWithMatchingPrefix(nsPrefix, out rDepth);
+            XmlAttribute? local = (XmlAttribute)nsLocallyDeclared[nsPrefix];
```
I'm like 99% sure that a null never goes into `nsLocallyDeclared`, meaning a null can't come out.

SteveDunn replied:
Thank you. I marked that as nullable because line 31 checks if for null.  I've marked it as non-nullable and left line 31 alone, as I understand that corrections like this is out of scope for these kind of PRs for nullable annotations.

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
More annotations

SteveDunn committed:
Remove red flag comment

SteveDunn committed:
PR feedback

SteveDunn committed:
Remove 'red flag' comments

SteveDunn commented:
> If this is not ready for review yet, please make this PR an actual draft. There should be a button to do that:
> 
> ![image](https://user-images.githubusercontent.com/6609929/160279875-40189bf1-0bc0-468a-ae65-05be34c3727e.png)

I might be missing something obvious, but I can't seem to find the 'not draft' button!  This all now compiles, and the tests all run, so it's no longer draft.

bartonjs commented on src/libraries/System.Runtime.InteropServices/gen/LibraryImportGenerator/Analyzers/ConvertToLibraryImportFixer.cs:
```diff
@@ -70,9 +70,9 @@ public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+                if (!bool.Parse(diagnostic.Properties[ConvertToLibraryImportAnalyzer.ExactSpelling]!))
                 {
-                    CharSet charSet = (CharSet)Enum.Parse(typeof(CharSet), diagnostic.Properties[ConvertToLibraryImportAnalyzer.CharSet]);
+                    CharSet charSet = (CharSet)Enum.Parse(typeof(CharSet), diagnostic.Properties[ConvertToLibraryImportAnalyzer.CharSet]!);
```
A lot of files in this PR don't look related, and have merge conflicts.  Presumably there's a misplaced commit in here somewhere.

SteveDunn replied:
Ah, yes. I think what's happened is that I looked at nullability 'solution wide'. I'll revert these ones

SteveDunn replied:
I've reverted everything that is not directly related `System.Security.Cryptograph.Xml`. Sorry about that. I'm used to having solutions where everything in it is related and under the same folder as the solution.

SteveDunn committed:
Revert unintended changes to tests

SteveDunn committed:
Revert unintended changes

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/C14NAncestralNamespaceContextManager.cs:
```diff
@@ -84,14 +84,14 @@ internal override void GetNamespacesToRender(XmlElement element, SortedList attr
+        internal override void TrackNamespaceNode(XmlAttribute? attr, SortedList nsListToRender, Hashtable nsLocallyDeclared)
         {
-            nsLocallyDeclared.Add(Utils.GetNamespacePrefix(attr), attr);
+            nsLocallyDeclared.Add(Utils.GetNamespacePrefix(attr!), attr);
```
Why are you sure that `attr` is not `null`, is it's explicitly marked as nullable?

SteveDunn replied:
Good point. `TrackNamespaceNode(XmlAttribute? attr...` doesn't the nullable.

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlNodeList.cs:
```diff
@@ -31,7 +31,7 @@ public override int Count
 
         // IList methods
-        public int Add(object value)
+        public int Add(object? value)
```
Is `null` ever gets added into this list? If no, it makes sense to assume it cannot contain `null`s, and annotate is accordingly (since this collection is internal). This will help reduce the number of `!`s in code that uses it

maxkoshevoi replied:
Oh, it implements `IList`. Missed that. Than there's no way around it I suppose..

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CipherData.cs:
```diff
@@ -63,7 +63,7 @@ public byte[] CipherValue
         public XmlElement GetXml()
         {
-            if (CacheValid) return _cachedXml;
+            if (CacheValid) return _cachedXml!;
```
Mark `CacheValid` with `[MemberNotNullWhen(true, nameof(_cachedXml))]`

SteveDunn replied:
Ah, yes, that's the attribute I was looking for. I was aware that there was a `NotNullWhen...` attribute.  I'll do this shortly, thanks for the info!

SteveDunn replied:
Great! I've updated this (and other types that have this pattern). One thing though: should these attributes also end up in the `ref` assembly? When I run `dotnet msbuild /t:GenerateReferenceAssemblySource`, it regenerates the file, but further builds result in compilation errors.

maxkoshevoi replied:
Yes, attributes and nullable annotations should be reflected in `ref`. 
I'd been adding all of those manually, but in theory nullable annotations should be updated in `ref` after running `dotnet msbuild /t:GenerateReferenceAssemblySource`

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/DataObject.cs:
```diff
@@ -97,7 +97,7 @@ private bool CacheValid
-        public XmlElement GetXml()
+        public XmlElement? GetXml()
         {
             if (CacheValid) return (_cachedXml);
```
Mark `CacheValid` with `[MemberNotNullWhen(true, nameof(_cachedXml))]`
And then `GetXml` might have non-nullable return type

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:
```diff
@@ -100,7 +101,7 @@ public SignedXml(XmlElement elem!!)
         }
 
-        private void Initialize(XmlElement element)
+        private void Initialize(XmlElement? element)
```
```suggestion
        [MemberNotNull(nameof(_containingDocument))]
        [MemberNotNull(nameof(m_signature))]
        [MemberNotNull(nameof(_safeCanonicalizationMethods))]
        private void Initialize(XmlElement? element)
```

And then `m_signature` (and others) can be marked as non-nullable, and `!`s can be removed.

SteveDunn replied:
Great suggestion, thanks! Those attributes are great.  However, it doesn't work for `_containingDocument` because:
```csharp
        private void Initialize(XmlElement? element)
        {
            _containingDocument = (element == null ? null : element.OwnerDocument);
```

So, `_containingDocument` can be `null` of `element` is null.

SteveDunn committed:
PR feedback: remove redundant !'s from things that can never return null

SteveDunn committed:
Added  `[MemberNotNullWhen(true, nameof(_cachedXml))]` on implementation.

SteveDunn committed:
Updated ref t ypes (with `dotnet msbuild /t:GenerateReferenceAssemblySource`)

SteveDunn committed:
Updated ref assembly

SteveDunn commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/ReferenceList.cs:
```diff
@@ -24,38 +24,42 @@ public int Count
 
-        public int Add(object value!!)
+#pragma warning disable CS8995 // Nullable type 'object?' is null-checked and will throw if null.
+        public int Add(object? value!!)
```
Ah, yes, this is an interesting one!  `Add` implements `Add` from `IList`. `IList` says `Add(object? value)`.
So, this signature must match `IList`, but when it does, then a `CS8995` compilation error occurs!

So, either way, there's a compilation error:
* without the
`error CS8767: Nullability of reference types in type of parameter 'value' of 'int ReferenceList.Add(object value)' doesn't match implicitly implemented member 'int IList.Add(object? value)' (possibly because of nullability attributes).`

* without the supression:
`error CS8995: Nullable type 'object?' is null-checked and will throw if null`

maxkoshevoi replied:
`IList` again..

SteveDunn committed:
PR feedback: add `MemberNotNull` attribute on `Initialize`)

SteveDunn committed:
Updated ref again

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -80,6 +80,7 @@ public abstract partial class EncryptedReference
         protected EncryptedReference() { }
         protected EncryptedReference(string? uri) { }
         protected EncryptedReference(string? uri, System.Security.Cryptography.Xml.TransformChain transformChain) { }
+        [System.Diagnostics.CodeAnalysis.MemberNotNullWhenAttribute(true, (string)null!)]
```
Why is second argument `null`? You can hardcode private member

SteveDunn replied:
I got an error when I used an actual name (or any non null value). Plus, I thought that null (as in, completely empty implementations) is the way to do it in refs?

bartonjs replied:
`MemberNotNull` attributes shouldn't be present in the ref.cs unless they're describing a public or protected field.

SteveDunn replied:
`_cachedXml` is `internal`. I removed the attribute from this file, but got the following compilation error:

` error : CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.MemberNotNullWhenAttribute' exists on 'System.Security.Cryptography.Xml.EncryptedReference.CacheValid.get()' in the implementation but not the reference`.

So I took a educated guess from that error message and put the attribute on the getter instead of the overall property, and there are no compilation errors.

maxkoshevoi replied:
Hm, why does this `ref` contains internal members at all? (`CacheValid`)

bartonjs replied:
Even on the getter, it's not an attribute we want in the ref.cs file, since it's not describing anything that can be observed by a caller.  If everything is building except this one item failing apicompat, you should be able to build like `dotnet msbuild /p:UpdateBaselineFiles=true` and it should add an exception for this item in the baseline file (which it might have to create).

> Hm, why does this ref contains internal members at all? (CacheValid)

`CacheValid` is `protected internal` which is `protected OR internal`.  I'm not 100% positive why we don't transform that to `protected` in the ref.cs files, but... we don't :)

SteveDunn replied:
I remove the attribute and ran `msbuild /p:UpdateBaselineFiles=true`. The errors from earlier appeared again. So I did a `dotnet clean`, and ran it again, and no errors! \o/  Thanks for your help.

SteveDunn replied:
@bartonjs - sorry, I'm really stumped on this one. I'm getting intermittent build errors. I can run `msbuild /p:UpdateBaselineFiles=true` one time, and get build errors, and run it again, and no build errors, and then run `dotnet build` and get build errors. I did try to research what's going on, but I can't find any documentation on `UpdateBaselineFiles`.  Any advice?  Currently, my branch in the local repo is identical to the branch on remote repo. My local branch builds fine, but I can see build errors in the PR.

SteveDunn replied:
Just to be clear that I'm doing the right thing: `CacheValid` **should not** be in the ref files? Or it **should** be in the ref files, but just listed as `protected` rather than `protected internal`? When I remove it from the ref file, everything builds fine.

SteveDunn replied:
Added back. I still get the build errors, but I seem to get them regardless of what's in the file.

SteveDunn committed:
Put `MemberNotNull` attribute on the getter rather than the property in the reference file.

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/C14NAncestralNamespaceContextManager.cs:
```diff
@@ -89,9 +89,9 @@ internal override void TrackNamespaceNode(XmlAttribute attr, SortedList nsListTo
 
-        internal override void TrackXmlNamespaceNode(XmlAttribute attr, SortedList nsListToRender, SortedList attrListToRender, Hashtable nsLocallyDeclared)
+        internal override void TrackXmlNamespaceNode(XmlAttribute? attr, SortedList nsListToRender, SortedList attrListToRender, Hashtable nsLocallyDeclared)
         {
-            nsLocallyDeclared.Add(Utils.GetNamespacePrefix(attr), attr);
+            nsLocallyDeclared.Add(Utils.GetNamespacePrefix(attr!), attr);
```
Here `attr` shouldn't be nullable too

SteveDunn replied:
The override forces us to make the parameter nullable. But `GetNamespacePrefix` doesn't expect a null. We can't change the signature of the base class, nor the signature of `GetNamespacePrefix`. I can't see an elegant way around this without changing anything outside the remit of these PRs.

maxkoshevoi replied:
> We can't change the signature of the base class
Why so? `AncestralNamespaceContextManager` is within this project, so you can change it's annotations

SteveDunn replied:
Very valid point. I just double checked and the base method is non null. Well spotted! Thanks.

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,54 +61,54 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
 
-        public override XmlElement CreateElement(string prefix, string localName, string namespaceURI)
+        public override XmlElement CreateElement(string? prefix, string localName, string? namespaceURI)
         {
-            return new CanonicalXmlElement(prefix, localName, namespaceURI, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlElement(prefix!, localName, namespaceURI!, this, _defaultNodeSetInclusionState);
```
Here `prefix` shouldn't be nullable

SteveDunn replied:
This is another of those situations where this method overrides the base method and so has to have the same signature. The nullable parameters here are incompatible with the constructor, which disallows nulls. It disallows nulls because the constructor calls straight into `base` which also disallows nulls.

I can't see a way to override the, er, override, and specify different annotations.

maxkoshevoi replied:
Hm, yes, too late to change annotations in `XmlDocument`, but why `namespaceURI` is non-nullable in `CanonicalXmlElement`? It is nullable in `XmlElement`'s ctor (where you are passing it to)

SteveDunn replied:
All of the callers to `CreateElement` pass a non-null `string`, so I chose the overload that takes a non-nullable string. The nullability in this parameter is only here because it needs to be for the overload, but in effect, the parameter is never null.

SteveDunn committed:
PR feedback: remove MemberNotNullWhenAttribute from ref

SteveDunn committed:
PR feedback: non nullable param to match method in base class

SteveDunn committed:
Fix post-merge build errors

SteveDunn commented:
I've just tidied up a few build errors after the recent rebase from main (after all the `!!`'s were removed).

I'm getting strange errors such as:
```
1>C:\git\runtime\src\libraries\Common\src\Interop\Windows\Advapi32\Interop.CryptAcquireContext.cs(25,36,25,55): error CA1420: Setting SetLastError to 'true' requires runtime marshalling to be enabled
1>C:\git\runtime\src\libraries\Common\src\Interop\Windows\Advapi32\Interop.CryptAcquireContext.cs(26,32,26,38): error CA1420: By-ref parameters require runtime marshalling to be enabled
````

Any ideas?

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -11,43 +11,50 @@ public sealed partial class CipherData
         public override System.Xml.XmlElement GetXml() { throw null; }
         public override void LoadXml(System.Xml.XmlElement value) { }
     }
+    public partial class CryptoSignedXmlRecursionException : System.Xml.XmlException
```
This type was declared `internal` in .NET Framework.  I don't know why it's `public` in our implementation here.  If the build fails with it declared internal in src we should just keep pruning it out of the ref.cs.

SteveDunn replied:
Thanks. I pruned it!

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/C14NAncestralNamespaceContextManager.cs:
```diff
@@ -26,8 +26,8 @@ private void GetNamespaceToRender(string nsPrefix, SortedList attrListToRender,
             int rDepth;
-            XmlAttribute local = (XmlAttribute)nsLocallyDeclared[nsPrefix];
-            XmlAttribute rAncestral = GetNearestRenderedNamespaceWithMatchingPrefix(nsPrefix, out rDepth);
+            XmlAttribute local = (XmlAttribute)nsLocallyDeclared[nsPrefix]!;
```
Since `local` gets checked for null on line 31 I don't think that non-null and the bang suppression are correct.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXml.cs:
```diff
@@ -117,11 +118,11 @@ internal byte[] GetBytes()
             _c14nDoc.WriteHash(hash, DocPosition.BeforeRootElement, _ancMgr);
             hash.TransformFinalBlock(Array.Empty<byte>(), 0, 0);
-            byte[] res = (byte[])hash.Hash.Clone();
+            byte[]? res = (byte[])hash.Hash!.Clone();
```
Clone only produces null if the input was null.  `hash.Hash` won't be null since TransformFinalBlock was called, and so Clone() won't return null.  We can get rid of some `?`s here and probably cascade a bit.

bartonjs replied:
Actually, this is the built-in `Array.Clone()`.  It can't ever return null, since it needed a non-null instance... so why is `res` declared as may-be null?  (I thought this was hitting an extension method that accepted null)

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CipherReference.cs:
```diff
@@ -41,7 +41,7 @@ internal byte[] CipherValue
         public override XmlElement GetXml()
         {
-            if (CacheValid) return _cachedXml;
+            if (CacheValid) return _cachedXml!;
```
Is this `!` still required now that CacheValid declares _cachedXml can't be null?

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedReference.cs:
```diff
@@ -1,33 +1,34 @@
         }
 
-        public string Uri
+        public string? Uri
```
Since the setter here throws for null, this property should be declared as

```suggestion
        [DisallowNull]
        public string? Uri
```

There are probably others in the same boat.

SteveDunn replied:
Done. But I get `error : CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.DisallowNullAttribute' exists on parameter 'value' on member 'System.Security.Cryptography.Xml.EncryptedReference.Uri.set(System.String)' in the implementation but not the reference.`

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedXml.cs:
```diff
@@ -204,46 +204,46 @@ private byte[] GetCipherValue(CipherData cipherData)
                         throw new CryptographicException(SR.Cryptography_Xml_UriNotSupported);
                     }
-                    decInputStream = tc.TransformToOctetStream(_document, _xmlResolver, baseUri);
+                    decInputStream = tc.TransformToOctetStream(_document!, _xmlResolver!, baseUri!);
```
TransformToOctetStream is internal.  If every caller is having to `!` away all their arguments, it seems easier to just declare it as null-accepting.

SteveDunn replied:
This is a tricky one. I'm concentrating on `baseUri`, e.g. when it's set on line 207. It's marked as nullable because there's a check for `_document == null ? null : _document.BaseURI`. In reality, `_document` is never null. So if that expression to check for null is removed, then `baseUri` could be declared as non-null. So that'd be one `!` we could get rid of to make the code easier to follow.
But, I was unsure if I should remove the `_document == null` part of the expression. What do you think?

SteveDunn replied:
But as far as the `_xmlResolver` field is concerned: it's explicitly set to null on construction, and can be mutated via the public `Resolver` property, but I cannot see anywhere in the type where it physically checked for null.
So, if I make the `resolver` parameter null-accepting, I then get more ambiguities further on, e.g on line 85 where we call `Utils.PreProcessStreamInput`, which accepts a non-null `resolver`. 

What I've done in light of this is make `xmlResolver`  null-accepting, including in the constructors of `CanonicalXml` (which seems to be the right thing to do, as there is an explicit null check for `inputStream` or `document`, but not for `resolver`).

The changes related to this are in their own commit for an easy revert if necessary.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/ExcCanonicalXml.cs:
```diff
@@ -65,11 +65,11 @@ internal byte[] GetBytes()
         }
 
-        internal byte[] GetDigestedBytes(HashAlgorithm hash)
+        internal byte[]? GetDigestedBytes(HashAlgorithm hash)
```
This can't ever return null.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Reference.cs:
```diff
@@ -352,20 +354,20 @@ internal byte[] CalculateHashValue(XmlDocument document, CanonicalXmlNodeList re
+            WebResponse? response = null;
+            Stream? inputStream = null;
+            XmlResolver? resolver = null;
+            byte[]? hashval = null;
```
hashval is only ever uninitialized or non-null.  It might work to declare it with no initializer, e.g.

```suggestion
            byte[] hashval;
```

Either way, this method can't return null.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXmlDebugLog.cs:
```diff
@@ -571,7 +571,7 @@ internal static void LogNamespacePropagation(SignedXml signedXml, XmlNodeList na
         /// <param name="data">Stream containing the output of the reference</param>
         /// <returns>Stream containing the output of the reference</returns>
-        internal static Stream LogReferenceData(Reference reference, Stream data)
+        internal static Stream? LogReferenceData(Reference reference, Stream? data)
```
I wonder if doing something like

```suggestion
        [return: NotNullIfNotNull("data")]
        internal static Stream? LogReferenceData(Reference reference, Stream? data)
```

will fix some unnecessary may-be-null states.

And since we'd NullReferenceException if `data` was null, maybe just getting rid of the `?`s here altogether is better.

SteveDunn replied:
Thank you!

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:
```diff
@@ -779,19 +782,19 @@ private static IList<string> DefaultSafeTransformMethods
         }
 
-        private byte[] GetC14NDigest(HashAlgorithm hash)
+        private byte[]? GetC14NDigest(HashAlgorithm hash)
```
Pretty sure this can't return null.

SteveDunn replied:
Line `789` says:
`if (isKeyedHashAlgorithm || !_bCacheValid || !SignedInfo!.CacheValid)`
In that `if` block is the only place where `_digestedSignedInfo` is set, so if the expression is `false`, then it'll return `null`.
But I'm not familiar with how this type works, so it could be that it's impossible for this expression to be `false` if it hasn't previously been `true`.
Looking at the usages of the return value, the consumers don't expect it to be `null`.
So I've changed it.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:
```diff
@@ -809,33 +812,33 @@ private byte[] GetC14NDigest(HashAlgorithm hash)
         private int GetReferenceLevel(int index, ArrayList references)
         {
-            if (_refProcessed[index]) return _refLevelCache[index];
+            if (_refProcessed![index]) return _refLevelCache![index];
```
A whole bunch of `!`s can be removed from this method by simply starting it with

```C#
Debug.Assert(_refProcessed != null);
Debug.Assert(_refLevelCache != null);
```

SteveDunn replied:
Thanks - I didn't realise that that acted as a hint to the null analyzer!

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Utils.cs:
```diff
@@ -396,7 +396,7 @@ internal static long Pump(Stream input, Stream output)
             long totalBytes = 0;
 
-            while ((numBytes = input.Read(bytes, 0, count)) > 0)
+            while ((numBytes = input!.Read(bytes, 0, count)) > 0)
```
I only see one caller to Pump.  Since we'll throw if `input` is null, it shouldn't be declared as null-accepting.  At worst, we're just moving this `!` to the caller where they're asserting it's not null.

Digging deeper, the `input` parameter can never be null (which is good, or we'd have a pending exception here).

SteveDunn replied:
Good spot. I've done some of the other suggestions above, and will do these when I get a spare minute later in the week. Thanks for your help!

SteveDunn replied:
I looked at the caller `EncrypedXml::GetCipherValue` and can see that all variations of `cipherData.CipherReference.Uri` are accounted for (is it null? - is it zero length? - is the first character a #? - otherwise throw) - which, as you say, means that it'll not be null. Have changed the input to reflect that.

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/Utils.cs:
```diff
@@ -756,9 +756,9 @@ internal static bool IsSelfSigned(X509Chain chain)
+        internal static AsymmetricAlgorithm? GetAnyPublicKey(X509Certificate2 certificate)
         {
-            AsymmetricAlgorithm algorithm = (AsymmetricAlgorithm)certificate.GetRSAPublicKey() ?? certificate.GetECDsaPublicKey();
+            AsymmetricAlgorithm? algorithm = (AsymmetricAlgorithm)certificate.GetRSAPublicKey()! ?? certificate.GetECDsaPublicKey();
```
Suffix-`!` followed by `??` makes no sense.  Perhaps the correct line is

```suggestion
            AsymmetricAlgorithm? algorithm = (AsymmetricAlgorithm?)certificate.GetRSAPublicKey() ?? certificate.GetECDsaPublicKey();
```

SteveDunn replied:
I don't know what I was thinking there!

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/XmlDecryptionTransform.cs:
```diff
@@ -150,8 +150,8 @@ private void LoadStreamInput(Stream stream)
             document.PreserveWhitespace = true;
-            XmlResolver resolver = (ResolverSet ? _xmlResolver : new XmlSecureResolver(new XmlUrlResolver(), BaseURI));
-            XmlReader xmlReader = Utils.PreProcessStreamInput(stream, resolver, BaseURI);
+            XmlResolver resolver = ((ResolverSet ? _xmlResolver : new XmlSecureResolver(new XmlUrlResolver(), BaseURI))!);
```
The extra parens don't seem necessary... the `!` could just be added outside the existing ones... or maybe just after the first clause?

```suggestion
            XmlResolver resolver = (ResolverSet ? _xmlResolver! : new XmlSecureResolver(new XmlUrlResolver(), BaseURI));
```

SteveDunn replied:
I went a step further and borrowed a technique I discovered earlier in the PR: I decorated `ResolverSet` with `[MemberNotNullWhen(true, nameof(_xmlResolver))]`
This meant that a lot of the extra parens and bangs melted away - which was very satisfying!

bartonjs commented:
> I'm getting strange errors...

Sounds like a bad incremental build state.  I recommend `git clean -xdf` and re-running build.sh/build.cmd.

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback from @bartonjs - add [DisallowNull] and cascade

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback

SteveDunn committed:
Fix build again, by removing the `protected internal` property `CacheValid`

bartonjs commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/TransformChain.cs:
```diff
@@ -150,12 +150,12 @@ public int Count
         }
 
-        internal Stream? TransformToOctetStream(XmlDocument document, XmlResolver resolver, string baseUri)
+        internal Stream? TransformToOctetStream(XmlDocument document, XmlResolver? resolver, string baseUri)
         {
             return TransformToOctetStream(document, typeof(XmlDocument), resolver, baseUri);
         }
```
`TransformToOctetStream(object,Type,XmlResolver,string)` already accepts the object as being null.

This method doesn't do anything special with document, just passes it into the null-accepting `inputObject` parameter.

So this method could accept document as being null, and get rid of some `!`s and `#pragma`s.

SteveDunn replied:
Good spot! Addressed in [this commit](https://github.com/dotnet/runtime/pull/67198/commits/4b66823c3ec36765fbf57dc3853335d7f2ffc027).

SteveDunn committed:
PR feedback

SteveDunn committed:
Added back `CacheValid` based on PR comment

SteveDunn commented:
This is one of a few PRs laying dormant. Is there anything I need to do to progress them?

SteveDunn commented:
> This is one of a few of my PRs laying dormant. Is there anything I need to do to progress them?

@bartonjs - would you be able to help me with this error?  I've attempted the fixes proposed before, but I can't consistently get a successful build:
`CannotRemoveAttribute : Attribute 'System.Diagnostics.CodeAnalysis.MemberNotNullWhenAttribute' exists on 'System.Security.Cryptography.Xml.EncryptedReference.CacheValid.get()' in the implementation but not the reference.`

stephentoub commented:
> would you be able to help me with this error? I've attempted the fixes proposed before, but I cannot consistently get a [successful build](https://dev.azure.com/dnceng/public/_build/results?buildId=1749019&view=results)

I think the problem is MemberNotNullWhen is missing from:
https://github.com/dotnet/runtime/blob/e18a77feff92fe6cdf8b891aa3b44fb1acb972d1/eng/DefaultGenApiDocIds.txt#L5-L9
We should add it there.

krwq commented:
@stephentoub @ViktorHofer was this PR meant to be closed?

ViktorHofer commented:
Nope. Thanks for pointing out.

SteveDunn committed:
Fix issues from recent rebase

SteveDunn committed:
Add `DisallowNull` attribute to ref file and remove it from `DefaultGenApiDocIds.txt`

SteveDunn commented:
@bartonjs @maxkoshevoi - I think everything is now addressed in the PR. Would you both agree?

SteveDunn commented:
Might it be possible for this be closed soon? I'm in a situation where I have a few PRs that have been ready to close for some time, and I'm needing to periodically rebase

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,54 +61,56 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+        public override XmlCDataSection CreateCDataSection(string? data)
         {
-            return new CanonicalXmlCDataSection(data, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlCDataSection(data!, this, _defaultNodeSetInclusionState);
```
Why are these `!`s necessary?  It looks like that first argument to all these CanonicalXx ctors is just passed down to the base constructor, which is annotated as `string?`... should all of these be `string?` as well?

SteveDunn replied:
`XmlDocument` declares this as `public virtual XmlCDataSection CreateCDataSection(string? data)`, so, from what I can see, this override must have the same signature.

The constructor for `CanonicalXmlCDataSection` looks like this:
```csharp
public CanonicalXmlCDataSection(string data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
```

So, it looks like the `!`s are necessary (at least to me, I stand to be corrected by you shortly :)  )

SteveDunn replied:
@stephentoub - just directly pinging you as these comments are not appearing on the 'conversation' tab. I think this is because I had to rebase due to changes on main.

stephentoub replied:
I'm not following. XmlCDataSection has this:
```C#
        protected internal XmlCDataSection(string? data, XmlDocument doc) : base(data, doc)
        {
        }
```
CanonicalXmlCDataSection has this:
```
        public CanonicalXmlCDataSection(string data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
        {
            _isInNodeSet = defaultNodeSetInclusionState;
        }
```
and this PR is providing the nullable annotations.  So shouldn't this PR be making it:
```C#
        public CanonicalXmlCDataSection(string? data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
        {
            _isInNodeSet = defaultNodeSetInclusionState;
        }
```
at which point the `!` on the first arugment to that ctor isn't needed?

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlElement.cs:
```diff
@@ -68,11 +68,11 @@ public void Write(StringBuilder strBuilder, DocPosition docPos, AncestralNamespa
                 foreach (object attr in attrListToRender.GetKeyList())
                 {
-                    (attr as CanonicalXmlAttribute).Write(strBuilder, docPos, anc);
+                    (attr as CanonicalXmlAttribute)!.Write(strBuilder, docPos, anc);
```
I think it'd be better to fail with a cast exception rather than with a null reference exception in the event that attr is something other than the specified type.  So instead of using `!`, the foreach can be changed to be e.g.
```C#
foreach (CanonicalXmlAttribute attr in nsListToRender.GetKeyList())
```
and then just use `attr` in the loop, rather than the `as` cast and suppression.  Same for other cases above and below.

SteveDunn replied:
Yes, I think so to, however I didn't change it because I thought it would change the external behaviour, which isn't encouraged in these types of PR (well, in general too, but _especially_ in these types of PRs).

SteveDunn replied:
@stephentoub - you may have missed the above comment; were my assumptions incorrect?

stephentoub replied:
Thanks.  We can leave it for now and look at cleaning it up subsequently.

SteveDunn replied:
OK. Should I mark this as Resolved? Or should an Issue be created to track it?

SteveDunn committed:
PR feedback

SteveDunn committed:
Fix issues from recent merge

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,52 +61,54 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+#pragma warning restore CS8765
         {
-            return new CanonicalXmlElement(prefix, localName, namespaceURI, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlElement(prefix!, localName, namespaceURI!, this, _defaultNodeSetInclusionState);
```
Note my previous comments about annotating constructors on these Canonical types and avoiding these `!`s apply to all of these cases, too, not just CanonicalXmlCDataSection.

SteveDunn replied:
Yes, you're right. I spotted those shortly after committing. I've checked the others and I _think_ there's no more.

SteveDunn committed:
Remove some more extraneous damnit operators

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.csproj:
```diff
@@ -1,7 +1,7 @@
   <PropertyGroup>
     <TargetFrameworks>$(NetCoreAppCurrent);$(NetCoreAppMinimum);netstandard2.0;$(NetFrameworkMinimum)</TargetFrameworks>
-    <Nullable>disable</Nullable>
+    <Nullable>enable</Nullable>
```
enable is the default now
```suggestion
```

maxkoshevoi replied:
Why is this resolved? `<Nullable>enable</Nullable>` line is still yet to be removed

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System.Security.Cryptography.Xml.csproj:
```diff
@@ -4,6 +4,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <Nullable>disable</Nullable>
     <NoWarn>$(NoWarn);CA1850</NoWarn> <!-- CA1850 suppressed due to multitargeting -->
+    <Nullable>enable</Nullable>
```
```suggestion
```

SteveDunn replied:
Removed. I also just noticed that there were **two** `<Nullable>` elements - probably a slip up when rebasing (there's been a few on this PR!)

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,52 +61,54 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
 
+#pragma warning disable CS8765 // Nullability doesn't match overriden member
         public override XmlElement CreateElement(string prefix, string localName, string namespaceURI)
+#pragma warning restore CS8765
```
Why is this being suppressed rather than making `CreateElement` match the base (which presumably would mean making prefix and namespaceURI both nullable)?  If the issue is CanonicalXmlElement's constructor, I'd expect it should also have prefix and namespaceURI as nullable, and if the issue in doing that is XmlElement's protected internal ctor having prefix as non-nullable, that looks like a bug that should be fixed: I don't see anything preventing it from being nullable.

SteveDunn replied:
That is exactly the problem. I will create a bug report and PR.

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,52 +61,54 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+        public override XmlComment CreateComment(string? data)
         {
-            return new CanonicalXmlComment(data, this, _defaultNodeSetInclusionState, _includeComments);
+            return new CanonicalXmlComment(data!, this, _defaultNodeSetInclusionState, _includeComments);
```
Why is this being `!`'d?  CanonicalXmlComment's ctor should have a `string? comment` parameter.

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlElement.cs:
```diff
@@ -30,12 +30,12 @@ public void Write(StringBuilder strBuilder, DocPosition docPos, AncestralNamespa
                 foreach (XmlAttribute attr in attrList)
                 {
-                    if (((CanonicalXmlAttribute)attr).IsInNodeSet || Utils.IsNamespaceNode(attr) || Utils.IsXmlNamespaceNode(attr))
+                    if ((((CanonicalXmlAttribute)attr)).IsInNodeSet || Utils.IsNamespaceNode(attr) || Utils.IsXmlNamespaceNode(attr))
```
Why the extra parens?

SteveDunn replied:
Well spotted - I can only think it's a result of fat-fingering

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CipherData.cs:
```diff
@@ -31,7 +33,7 @@ private bool CacheValid
         }
 
-        public CipherReference CipherReference
+        public CipherReference? CipherReference
```
This doesn't look right.  The setter is explicitly throwing an exception if value is null.  Same for CipherValue below.

SteveDunn replied:
Yes, I missed the fact that the constructors call either one of two setters, each of which guarantee non-null.

SteveDunn replied:
Actually, there is a default constructor too. I could be wrong, but I think these **do** need to be nullable...

stephentoub replied:
Making them nullable implies you can set them to null, but the setter explicitly throws for null.  If it's going to remain `public CipherReference? CipherReference`, then it also needs `[DisallowNull]`.

SteveDunn replied:
But somebody _could_ do `new CipherData().CipherReference.ToString()` and get an NRE.  That would be pretty silly though, so could you clarify if we should be annotating for every concievable usage, or just for 'reasonable' usage. (or perhaps I've missed something completely...?)

stephentoub replied:
We're talking past each other. It should be:
```C#
[DisallowNull]
public CipherReference? CipherReference
```
which says the property may return null but can't be set to null.

SteveDunn replied:
I see. Thank you.

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedReference.cs:
```diff
@@ -1,16 +1,17 @@
-        private string _referenceType;
-        private TransformChain _transformChain;
-        internal XmlElement _cachedXml;
+        private string? _uri;
```
In what situation will this be null?

SteveDunn replied:
It looks like it can't, so I have declared it as `private string _uri = null!;`. My other option was this:

![image](https://user-images.githubusercontent.com/263416/178093171-a6bf84df-36d0-4f0a-9577-dd2d0ce87b83.png)

But it still required `!`ing `_uri` from the getter

SteveDunn replied:
I have a suspicion that maybe there's an attribute that I haven't thought of... So I'll leave this one as unresolved for the time being.

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedReference.cs:
```diff
@@ -27,7 +28,8 @@ protected EncryptedReference(string uri, TransformChain transformChain)
 
-        public string Uri
+        [DisallowNull]
+        public string? Uri
```
I don't see how _uri becomes null.  Shouldn't this just be `public string Uri` with no additional attributes?

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedXml.cs:
```diff
@@ -185,7 +185,7 @@ private byte[] GetCipherValue(CipherData cipherData)
             }
 
-            Stream inputStream = null;
+            Stream? inputStream = null;
```
I can't comment on it directly, but Recipient above also appears like it would benefit from an [AllowNull].  In general, if the setter allows a null to be set and the getter checks for null and returns non-null, the property can be non-nullable with [AllowNull].

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedXml.cs:
```diff
@@ -204,7 +204,7 @@ private byte[] GetCipherValue(CipherData cipherData)
                 {
                     // self referenced Uri
-                    string baseUri = _document?.BaseURI;
+                    string baseUri = _document?.BaseURI!;
```
This is a strange construction... if _document is null, baseUri will be null.  Why isn't it `string? baseUri`?  Maybe there are `?`s missing elsewhere?

SteveDunn replied:
I remember it seeming strange when typing it actually! `baseUri` can't be nullable, because it's the last parameter in `TransformChain::TransformToOctetStream`, which _isn't_ nullable.

But, digging through a bit more, this value ulimately ends up being stored in `Transform::BaseURI`, which **is** nullable.

I'll change the parameters to be nullable.

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedInfo.cs:
```diff
@@ -62,7 +63,7 @@ public object SyncRoot
         //
 
-        public string Id
+        public string? Id
```
CanonicalizationMethod below is another example of a property that should be [AllowNull]

stephentoub commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/SignedXml.cs:
```diff
@@ -171,7 +175,7 @@ public EncryptedXml EncryptedXml
             {
                 if (_exml == null)
-                    _exml = new EncryptedXml(_containingDocument); // default processing rules
+                    _exml = new EncryptedXml(_containingDocument!); // default processing rules
```
[AllowNull]

SteveDunn committed:
Fix NRT analysis errors after prior rebase

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback

SteveDunn commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,52 +61,54 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+            return new CanonicalXmlSignificantWhitespace(text!, this, _defaultNodeSetInclusionState);
         }
 
         public override XmlProcessingInstruction CreateProcessingInstruction(string target, string data)
```
I have created an [Issue](https://github.com/dotnet/runtime/issues/71862) for this one. I started on a PR, but it turned into a bit of a rabbit hole and I stopped when I reached `XmlNode`!

SteveDunn replied:
@stephentoub - I made the change in that other PR, and you merged it. I've rebased and rebuilt (everything), but for some reason, it's still picking up the signature of the old version (without the `?` on `XmlElement):
`Possible null reference argument for parameter 'prefix' in 'XmlElement.XmlElement(string prefix, string localName, string? namespaceURI, XmlDocument doc)`

Do I need to do anything special projects named `Special` (e.g. `System.Private.xml`)?

stephentoub replied:
In which build are you seeing the compilation errors?  This project builds multiple times against multiple frameworks:
https://github.com/dotnet/runtime/blob/cae954add0f01c85de7fef8cc41c2a11d1c920e1/src/libraries/System.Security.Cryptography.Xml/src/System.Security.Cryptography.Xml.csproj#L3
My assumption is the error is coming from the build targeting .NET 6, which won't have the updates to the System.Xml.ReaderWriter annotations, which are only in .NET 7.

SteveDunn replied:
Yes, that was it. Added a conditional compilation.

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback

SteveDunn committed:
PR feedback

SteveDunn committed:
CipherReference and CipherValue made to disallow null

SteveDunn committed:
Make Uri non-nullable on EnctrypedReference

SteveDunn committed:
Add AllowNull attribute on EncryptedType::KeyInfo

SteveDunn committed:
AllowNull on EncryptedXml::Recipient

SteveDunn committed:
Made baseUri nullable

SteveDunn committed:
DisallowNull on PropertlyElement

SteveDunn committed:
AllowNull on SignedInfo::CanonicalizationMethod

SteveDunn committed:
AllowNull on SignedXml::EncryptedXml

SteveDunn committed:
Fix build after recent rebase

SteveDunn committed:
Remove redundant damnit operator now that https://github.com/dotnet/runtime/pull/71860 has been merged

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedReference.cs:
```diff
@@ -1,16 +1,17 @@
-        private string _referenceType;
-        private TransformChain _transformChain;
-        internal XmlElement _cachedXml;
+        private string _uri = null!;
```
Mark `Uri` property's setter with `[MemberNotNull(nameof(_uri))]`, and you won't need to do ` = null!`

SteveDunn replied:
Thank you @maxkoshevoi !

maxkoshevoi commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/TransformChain.cs:
```diff
@@ -52,23 +52,23 @@ public int Count
         // The goal behind this method is to pump the input stream through the transforms and get back something that
         // can be hashed
-        internal Stream TransformToOctetStream(object inputObject, Type inputType, XmlResolver resolver, string baseUri)
+        internal Stream? TransformToOctetStream(object? inputObject, Type inputType, XmlResolver? resolver, string? baseUri)
```
When will this method ever return null? 
In two cases it just returns `new MemoryStream`, 
and in one it returns `if (currentInput is Stream) return currentInput as Stream;` (which is also never null)

Also upstream calls of this method (in `EncryptedXml.cs`) don't expect it to return null

stephentoub replied:
```C#
if (currentInput is Stream inputStream)
{
    return inputStream;
}
```

SteveDunn replied:
Of course! It's been a long day...

SteveDunn committed:
PR feedback

SteveDunn committed:
Fix build due to change of the CanonicalXmlElement constructor

SteveDunn committed:
PR feedback

SteveDunn commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedXml.cs:
```diff
@@ -229,14 +231,14 @@ private byte[] GetCipherValue(CipherData cipherData)
                 using (MemoryStream ms = new MemoryStream())
                 {
-                    Utils.Pump(decInputStream, ms);
+                    Utils.Pump(decInputStream!, ms);
```
Well spotted! Removed these, and a couple of others found by ReSharper

SteveDunn committed:
Remove redundant damnit operators

danmoseley commented:
@SteveDunn this appears to be breaking some tests. could you please look?

SteveDunn committed:
Fix broken tests

SteveDunn commented:
> @SteveDunn this appears to be breaking some tests. could you please look?

They look ok now @danmoseley . Looks related to a merge issue during a rebase.

danmoseley commented:
@SteveDunn thanks. i fixed a small merge conflict. is next action here on us - is all feedback addressed?

SteveDunn commented:
> @SteveDunn thanks. i fixed a small merge conflict. is next action here on us - is all feedback addressed?

@danmoseley , I'm fairly sure everything has been addressed;  I've fixed things as comments have arrived. It seems like System.Xml might need a few annotation tweaks in the future, which might benefit the accuracy of annotations here.

stephentoub commented:
Someone who understands the System.Security.Cryptography.Xml surface area should do a pass through the ref .cs to make sure the annotations make sense.  Is that you, @bartonjs? Or maybe @vcsjones if you have spare cycles?

vcsjones commented:
> maybe @vcsjones if you have spare cycles?

I can look at this sometime later this week if no one beats me to it.

SteveDunn commented:
Is anyone free to look at this one? Would be nice to get it closed and merged.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,42 +61,42 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+        public override XmlWhitespace CreateWhitespace(string? prefix)
         {
-            return new CanonicalXmlWhitespace(prefix, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlWhitespace(prefix!, this, _defaultNodeSetInclusionState);
```
this looks like potential NRE, might be good to open an issue (don't fix the product in this PR)

SteveDunn replied:
It looks like `CanonicalXmlWhitespace` takes a nullable `prefix`. Have remove the bang from this line.

SteveDunn replied:
`CanonicalXmlWhitespace` now takes a nullable `prefix`

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,42 +61,42 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
+        public override XmlSignificantWhitespace CreateSignificantWhitespace(string? text)
         {
-            return new CanonicalXmlSignificantWhitespace(text, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlSignificantWhitespace(text!, this, _defaultNodeSetInclusionState);
```
ditto

SteveDunn replied:
`CanonicalXmlSignificantWhitespace` now takes a nullable `text` - have removed th bang.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs:
```diff
@@ -61,42 +61,42 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
 
         public override XmlProcessingInstruction CreateProcessingInstruction(string target, string data)
         {
             return new CanonicalXmlProcessingInstruction(target, data, this, _defaultNodeSetInclusionState);
```
I've changed this in the other PR, `data` is nullable now

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlElement.cs:
```diff
@@ -12,8 +12,12 @@ internal sealed class CanonicalXmlElement : XmlElement, ICanonicalizableNode
         private bool _isInNodeSet;
 
-        public CanonicalXmlElement(string prefix, string localName, string namespaceURI, XmlDocument doc, bool defaultNodeSetInclusionState)
+        public CanonicalXmlElement(string? prefix, string localName, string? namespaceURI, XmlDocument doc, bool defaultNodeSetInclusionState)
+#if NET7_0_OR_GREATER
             : base(prefix, localName, namespaceURI, doc)
+#else
+            : base(prefix!, localName, namespaceURI, doc)
+#endif
```
honestly, probably better to always add `!` than have conditional compile here. It will result in the same IL in the end

stephentoub replied:
Yes,  we shouldn't use conditionals just to avoid a `! `. `! ` has zero impact on the IL generated.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlNodeList.cs:
```diff
@@ -80,7 +80,7 @@ public bool IsReadOnly
         }
 
-        object IList.this[int index]
+        object? IList.this[int index]
```
is it actually possible for this to return null? we're throwing when user tries to add null

SteveDunn replied:
It doesn't look like it, but it must match the index operator for `IList`

krwq replied:
I don't expect this to be fixed, just unresolving as FYI:
for return values you can make it non-nullable even if interface returns nullable (for setter it has to match but you can make it non-nullable with `[AllowNull]`)

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlNodeList.cs:
```diff
@@ -43,24 +43,24 @@ public void Clear()
         }
 
-        public void Insert(int index, object value)
+        public void Insert(int index, object? value)
```
I assume the `?` is here because of the interface implementation? the getters can still be non-null though because that doesn't break interface

SteveDunn replied:
Yes, it's to satisfy the interface. `Item` is already non-null.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CipherData.cs:
```diff
@@ -1,15 +1,16 @@
-        private byte[] _cipherValue;
+        private XmlElement? _cachedXml;
+        private CipherReference? _cipherReference;
+        private byte[]? _cipherValue;
```
since you annotated constructor as non-nullable I'd expect this to be non-null and all related members, or alternatively ctor to accept null but looking at the existing setter I'd assume this is just missing null check in the ctor

SteveDunn replied:
It's nullable as there are multiple constructors, not all of which set it.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/DSAKeyValue.cs:
```diff
@@ -23,7 +23,7 @@ public DSAKeyValue()
         }
 
-        public DSAKeyValue(DSA key)
+        public DSAKeyValue(DSA? key)
```
this might be worth filing an issue (or adding to existing one) since we accept null here but later throw if it's null. Unless any of the code assumes this can be null I'd go ahead and change annotation to non-null to avoid breaking change in the future

SteveDunn replied:
Good point. Line 81 would throw an NRE, so I've changed the annotation to be non null as suggested.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/DSASignatureDescription.cs:
```diff
@@ -19,15 +19,15 @@ public DSASignatureDescription()
         public sealed override AsymmetricSignatureDeformatter CreateDeformatter(AsymmetricAlgorithm key)
         {
-            var item = (AsymmetricSignatureDeformatter)CryptoConfig.CreateFromName(DeformatterAlgorithm);
+            var item = (AsymmetricSignatureDeformatter)CryptoConfig.CreateFromName(DeformatterAlgorithm!)!;
```
seems to me that wither `DeformatterAlgorithm` or `CreateFromName` is badly annotated or we're missing checks somewhere here

SteveDunn replied:
`DeformatterAlgorithm` looks correct - it has a public setter that doesn't check for the value being assinged.

`CreateFromName` also **can** return null.

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/EncryptedReference.cs:
```diff
@@ -30,6 +31,8 @@ protected EncryptedReference(string uri, TransformChain transformChain)
         {
             get { return _uri; }
+
+            [MemberNotNull(nameof(_uri))]
```
nit: that seems like rather weird annotation to have on the setter, I'd recommend avoiding it (feel free to ignore this assuming it doesn't flow to the ref)

SteveDunn replied:
I'm pretty sure this was the result of other PR feedback. I've lost the history because of all the rebases required because of the age of this PR.

krwq commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -11,8 +11,10 @@ public sealed partial class CipherData
+        [System.Diagnostics.CodeAnalysis.DisallowNullAttribute]
+        public System.Security.Cryptography.Xml.CipherReference? CipherReference { get { throw null; } set { } }
+        [System.Diagnostics.CodeAnalysis.DisallowNullAttribute]
+        public byte[]? CipherValue { get { throw null; } set { } }
```
double check this can be null given constructor annotation. if none of the code expects it might be worth filing an issue on ctor not throwing (same for CipherReference)

bartonjs replied:
That one's easy: `byte[] ret = new CipherData().CipherValue;`, `ret` is `null`.

krwq replied:
fair enough, I didn't notice parameterless ctor

krwq commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -46,8 +48,8 @@ public partial class DSAKeyValue : System.Security.Cryptography.Xml.KeyInfoClaus
         public DSAKeyValue() { }
-        public DSAKeyValue(System.Security.Cryptography.DSA key) { }
-        public System.Security.Cryptography.DSA Key { get { throw null; } set { } }
+        public DSAKeyValue(System.Security.Cryptography.DSA? key) { }
+        public System.Security.Cryptography.DSA? Key { get { throw null; } set { } }
```
if none of the code passes in null and all usages throw consider changing key to be non-null and filing an issue on lack of throwing

SteveDunn replied:
Removed nullability in both and will file an issue for missing null checks.

SteveDunn replied:
https://github.com/dotnet/runtime/issues/75048

krwq commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -60,8 +62,9 @@ public sealed partial class EncryptedData : System.Security.Cryptography.Xml.Enc
-        public string Recipient { get { throw null; } set { } }
+        public string? CarriedKeyName { get { throw null; } set { } }
+        [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
+        public string? Recipient { get { throw null; } set { } }
```
show this return non-null? if not I'm not sure I understand AllowNull here

SteveDunn replied:
This was the result of earlier PR feedback

stephentoub replied:
This combination doesn't make sense.  The AllowNull only makes sense if the property type is `string` rather than `string?`.

SteveDunn replied:
Missed this one. Changed to be non-null like the implementation.

krwq commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -173,7 +179,7 @@ public sealed partial class EncryptionPropertyCollection : System.Collections.IC
         public System.Security.Cryptography.Xml.EncryptionProperty this[int index] { get { throw null; } set { } }
         public object SyncRoot { get { throw null; } }
-        object System.Collections.IList.this[int index] { get { throw null; } set { } }
+        object? System.Collections.IList.this[int index] { get { throw null; } set { } }
```
is this actually nullable if we throw when null is added?

krwq replied:
same comment for every IList

SteveDunn replied:
As far as I know, it needs to match the `IList` signature.

krwq reviewed (approved):
LGTM after resolving comments and merge conflicts

SteveDunn committed:
Fix issues after rebase

SteveDunn commented:
> LGTM after resolving comments and merge conflicts

I've rebased and looked at the issues you've mentioned.

SteveDunn committed:
PR feedback

krwq commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -60,8 +62,9 @@ public sealed partial class EncryptedData : System.Security.Cryptography.Xml.Enc
-        public string Recipient { get { throw null; } set { } }
+        public string? CarriedKeyName { get { throw null; } set { } }
+        [System.Diagnostics.CodeAnalysis.AllowNullAttribute]
+        public string? Recipient { get { throw null; } set { } }
```
This doesn't seem to be fixed: https://github.com/dotnet/runtime/pull/67198/files#r963817874

krwq commented on src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/DSAKeyValue.cs:
```diff
@@ -155,34 +155,34 @@ public override void LoadXml(XmlElement value)
             try
             {
-                Key.ImportParameters(new DSAParameters
+                Key!.ImportParameters(new DSAParameters
```
note these won't need `!` once this is fixed

SteveDunn replied:
Removed

krwq reviewed (approved):
looks good, there seem to be couple of unaddressed comments

SteveDunn commented on src/libraries/System.Security.Cryptography.Xml/ref/System.Security.Cryptography.Xml.cs:
```diff
@@ -47,7 +49,7 @@ public partial class DSAKeyValue : System.Security.Cryptography.Xml.KeyInfoClaus
         public DSAKeyValue() { }
         public DSAKeyValue(System.Security.Cryptography.DSA key) { }
-        public System.Security.Cryptography.DSA Key { get { throw null; } set { } }
+        public System.Security.Cryptography.DSA? Key { get { throw null; } set { } }
```
Sorry - it really is non-nullable now.

SteveDunn commented:
> Please do not rebase unless there are merge conflicts. Just push changes with those three fixes (or respond to the comments if not addressing for whatever reason)

I don't rebase unless there are merge conflicts. Unfortunately, last time, there were merge conflicts. I think I've done 4 or 5 rebases for this PR!

SteveDunn committed:
PR feedback

SteveDunn commented:
Hi - I think everything is now addressed here. Might we be able to merge it before a rebase is due again? :)

SteveDunn commented:
Hi again, can this be merged/closed before another rebase is required? Thanks in advance.

stephentoub commented:
@bartonjs, can you help land this?

krwq reviewed (approved):
Looks good from my side. Thank you!

krwq commented:
Contributes to: https://github.com/dotnet/runtime/issues/41720

stephentoub commented:
Thank you, @SteveDunn.
